src/reaction_network.jl

```diff
@@ -203,144 +257,196 @@ struct ReactionStruct
 end
 
+# Function for creating a Reaction structure (used by the @reaction macro).
 function make_reaction(ex::Expr)
 
-    # handle interpolation of variables
+    # Handle interpolation of variables
     ex = esc_dollars!(ex)
 
-    # parse DSL lines
+    # Parses reactions, species, and parameters.
     reaction = get_reaction(ex)
-    allspecies = get_reactants(reaction)   # species defined by stoich
-    parameters = get_rx_species([reaction], Symbol[])  # anything in a rate is a parameter
-    !isempty(intersect(forbidden_symbols, union(allspecies, parameters))) &&
-        error("The following symbol(s) are used as species or parameters: " *
-              ((map(s -> "'" * string(s) * "', ",
-                    intersect(forbidden_symbols, union(species, parameters)))...)) *
-              "this is not permited.")
-
-    pexprs = get_pexprs(parameters)    # parameters
-    sexprs = get_sexprs(allspecies)    # species
-    rxexpr = get_rxexprs(reaction)     # reaction
+    species, parameters = extract_species_and_parameters!([reaction], [])
 
+    # Checks for input errors.
+    forbidden_symbol_check(union(species, parameters))
+
+    # Creates expressions corresponding to actual code from the internal DSL representation.
+    sexprs = get_sexpr(species, Dict{Symbol, Expr}())
+    pexprs = get_pexpr(parameters, Dict{Symbol, Expr}())
+    rxexpr = get_rxexprs(reaction)
+
+    # Returns the rephrased expression.
     quote
         $pexprs
+        :(@variables t)
         $sexprs
         $rxexpr
     end
 end
 
-function get_rx_species(rxs, ps)
-    pset = Set(ps)
-    species_set = Set{Symbol}()
-    for rx in rxs
-        find_species_in_rate!(species_set, rx.rate, pset)
-        for sub in rx.substrates
-            find_species_in_rate!(species_set, sub.stoichiometry, pset)
-        end
-        for prod in rx.products
-            find_species_in_rate!(species_set, prod.stoichiometry, pset)
+### Auxiliary function called by the main make_reaction_system and make_reaction functions. ###
+
+# Function that handles variable interpolation.
+function esc_dollars!(ex)
+    if ex isa Expr
+        if ex.head == :$
+            return esc(:($(ex.args[1])))
+        else
+            for i in 1:length(ex.args)
+                ex.args[i] = esc_dollars!(ex.args[i])
+            end
         end
     end
-    collect(species_set)
+    ex
+end
+
+# When the user have used the @species (or @parameters) options, extract species (or
+# parameters) from its input.
+function extract_syms(ex::Expr, vartype::Symbol)
+    vars = Symbolics._parse_vars(vartype, Real, ex.args[3:end])
+    Vector{Union{Symbol, Expr}}(vars.args[end].args)
 end
 
-function find_species_in_rate!(sset, rateex::ExprValues, ps)
+# Function looping through all reactions, to find undeclared symbols (species or
+# parameters), and assign them to the right category.
+function extract_species_and_parameters!(reactions, excluded_syms)
+    species = OrderedSet{Union{Symbol, Expr}}()
```

Comment:
I changed this and parameters below to `OrderedSet`s. While for small networks this will likely be a bit slower, the DSL is so fast it shouldn't be noticeable. In contrast, I think this should help performance for larger networks. 

LMK if this is ok with you, or you had a reason to switch the older code to `Vector`s.

Reply:
See also the tweaks to `add_syms_from_expr!`.

Reply:
Think this should be good. If I remember next time we meet I will have to ask you why this improves stuff and I might learn something.

Reply:
To be honest, I would have preferred to use a `Set` but that is breaking lots of tests and I don't want to handle updating them in this PR. (`Set` at least used to have better performance than `OrderedSet`.)